### ISSUE TITLE
[regression] crash in crossref filter

### ISSUE THREAD

**Issue by kcarnold** [quarto-dev/quarto-cli#7259]
### Bug description

A colleague uses blocks with ids (see example below). This markup crashed the crossref filter:

```
Error running filter /Users/ka37/code/github.com/quarto-dev/quarto-cli/src/resources/filters/main.lua:
...ev/quarto-cli/src/resources/filters/./crossref/index.lua:202: attempt to index a nil value (field 'integer index')
stack traceback:
        ...ev/quarto-cli/src/resources/filters/./crossref/index.lua:100: in field 'Pandoc'
        ...v/quarto-cli/src/resources/filters/./ast/customnodes.lua:72: in function 'run_emulated_filter'
        .../quarto-cli/src/resources/filters/./ast/runemulation.lua:40: in local 'callback'
        .../quarto-cli/src/resources/filters/./ast/runemulation.lua:54: in upvalue 'run_emulated_filter_chain'
        .../quarto-cli/src/resources/filters/./ast/runemulation.lua:89: in function <.../quarto-cli/src/resources/filters/./ast/runemulation.lua:86>
ERROR: Error
    at renderFiles (file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/command/render/render-files.ts:334:23)
    at eventLoopTick (ext:core/01_core.js:181:11)
    at async renderProject (file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/command/render/project.ts:266:23)
    at async Command.fn (file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/command/preview/cmd.ts:292:36)
    at async Command.execute (file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/vendor/deno.land/x/cliffy@v0.25.4/command/command.ts:1790:7)
    at async quarto (file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/quarto.ts:135:3)
    at async file:///Users/ka37/code/github.com/quarto-dev/quarto-cli/src/quarto.ts:167:5
```

`git bisect` said that this is the first bad commit: e5faa8d74dd9f9b41505e1130ce913e9d9e38ba6

### Steps to reproduce

On main branch, compile the following:

````
---
title: "oops"
---

This is a test.

::: {#exr-a-b}
This is a named exercise.
:::
````

### Your environment

- OS: macOS Ventura 13.6 (22G120)

### Quarto check output

```bash
$ quarto check
Quarto 99.9.9
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.8: OK
      Dart Sass version 1.55.0: OK
      Deno version 1.33.4: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 99.9.9
      Path: /Users/ka37/code/github.com/quarto-dev/quarto-cli/package/dist/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /Library/TeX/texbin
      Version: 2023

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.11.6
      Path: /usr/local/opt/python@3.11/bin/python3.11
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.3.1
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Users/ka37/Library/R/4.2/library
        - /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library
      knitr: 1.44
      rmarkdown: 2.24

[✓] Checking Knitr engine render......OK
```

**Comment by cscheid** [comment]
Thanks for the report. I can't seem to reproduce:

````
$ cat exercise.qmd
---
title: "oops"
---

This is a test.

::: {#exr-a-b}
This is a named exercise.
:::
$ quarto render exercise.qmd
pandoc
  to: html
  output-file: exercise.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: oops

Output created: exercise.html
````

Since you're reporting `99.9.9`: `git log` shows that I'm on

````
commit 749a3527a8a79ccc1f601ce1f2b62dd2b90a7528 (HEAD -> main, origin/main, origin/HEAD)
Author: Christophe Dervieux <christophe.dervieux@gmail.com>
Date:   Tue Oct 17 14:24:40 2023 +0200

    Correctly quote and test for extension when space in filepath
````

**Comment by cscheid** [comment]
Oh, wait, "crossref filter". Hold on.

**Comment by cscheid** [comment]
I can repro this, thanks for the report.

**Comment by cscheid** [comment]
This should be closed on `main` now:

````
$ cat exercise.qmd
---
title: "oops"
---

This is a test.

::: {#exr-a-b}
This is a named exercise.
:::
$ quarto editor-support crossref --input exercise.qmd
{"entries":[{"key":"exr-a-b","caption":""}]}%
```

Thanks again for the report.

**Comment by kcarnold** [comment]
I still get the same crash. It looks like one code-path got fixed but there's almost identical code on a different path that wasn't changed:

https://github.com/quarto-dev/quarto-cli/blob/86810a58ab723c356fc7ff6fcf8bb83cb357ecbf/src/resources/filters/crossref/index.lua#L198

(Is there a meaningful difference between a crossref's `.caption` being `nil` vs being empty? There are several other checks against `nil` in that file that might warrant further review.)

**Comment by cscheid** [comment]
Very annoying - I'm sorry and thanks for the continued patience.  I'll fix right away.


> (Is there a meaningful difference between a crossref's `.caption` being `nil` vs being empty? There are several other checks against `nil` in that file that might warrant further review.)

Yes, there is a difference, from line 67 in that file. (I am now inclined to change that part of the code as well.)

Until this PR we had essentially no test coverage on IDE crossref resolution, so I was trying to be as conservative as possible. I now think that's a losing proposition and we should make the _good_ fix here.

**Comment by cscheid** [comment]
@kcarnold would you mind trying this on `main` now? Thanks!

**Comment by mcanouil** [comment]
It's now working/compiling but the output is not great for html (pdf is fine), *i.e.*, the parenthesis are there even with no labels.

<img width="576" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/2761d382-987e-464e-9326-40016845fc04">

````qmd
---
title: "Quarto Playground"
format:
  pdf: default
  html: default
number-sections: true
---

# Section

This is a playground for Quarto.

::: {#thm-line}
The equation of any straight line, called a linear equation, can be written as:

$$
y = mx + b
$$
:::

See @thm-line.


::: {#exr-a-b}
This is a named exercise.
:::
````

**Comment by mcanouil** [comment]
`docx` exhibits the sameb behaviour as `html`.
<img width="517" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/7e662012-b0d5-45ac-9b8c-53acf23f82c9">

I tried with Typst and it does not compile but do not know if it is suppose to handle `#thm-line`.
For `#exr` it compiles and produces the same kind of output as `docx` and `html`.
<img width="368" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/cd9a5aa3-a873-4067-88a3-cbdd3bea8fce">

**Comment by mcanouil** [comment]
A bit more comprehensive try:

````qmd
---
title: "Quarto Playground"
format:
  pdf: default
  html: default
number-sections: true
---

# Section

::: {#thm-example}
Something without label/name ...
:::

::: {#lem-example}
Something without label/name ...
:::

::: {#cor-example}
Something without label/name ...
:::

::: {#prp-example}
Something without label/name ...
:::

::: {#cnj-example}
Something without label/name ...
:::

::: {#def-example}
Something without label/name ...
:::

::: {#exm-example}
Something without label/name ...
:::

::: {#exr-example}
Something without label/name ...
:::

````

| HTML | LaTeX | Typst | Docx |
|--------|--------|--------|--------|
| <img width="331" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/1aa93c66-ee7f-4847-9d0f-d7121aadfbe7"> | <img width="352" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/e2dbb80b-0560-4944-a7e9-724cdce8fbd5"> | <img width="301" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/cb5b40de-c960-4741-a1be-bbb860723a5a"> | <img width="368" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/95446115-2f81-469c-ae78-90277c16bec1"> |

**Comment by cscheid** [comment]
Fix incoming on #7275.

**Comment by kcarnold** [comment]
I confirm this fixes the crash for me.  (I also see the formatting issue, but you're addressing that.)

Thanks!
